pkg/evaluation/evaluator.go

```diff
@@ -0,0 +1,355 @@
+			log.Debugf("Target %s excluded from segment %s via exclude list", target.Name, segment.Name)
+			return false
+		}
+
+		// Should Target be included - if in included list we return true
+		if isTargetInList(target, segment.Included) {
+			log.Debugf(
+				"Target %s included in segment %s via include list",
+				target.Name,
+				segment.Name)
+			return true
+		}
+
+		// Should Target be included via segment rules
+		rules := segment.Rules
+		if rules != nil && e.evaluateClauses(*rules, target) {
+			log.Debugf(
+				"Target %s included in segment %s via rules", target.Name, segment.Name)
+			return true
+		}
+	}
+	return true
+}
+
+func (e Evaluator) checkPreRequisite(parent *rest.FeatureConfig, target *rest.Target) bool {
+	prerequisites := parent.Prerequisites
+	if prerequisites != nil {
+		log.Infof(
+			"Checking pre requisites %v of parent feature %v",
+			prerequisites,
+			parent.Feature)
+		for _, pre := range *prerequisites {
+			prereqFeature := pre.Feature
+			prereqFeatureConfig, err := e.query.GetFlag(prereqFeature)
+			if err != nil {
+				log.Errorf(
+					"Could not retrieve the pre requisite details of feature flag : %v", prereqFeature)
+				return true
+			}
+
+			prereqEvaluatedVariation, err := e.evaluateFlag(prereqFeatureConfig, target)
+			if err != nil {
+				log.Errorf(
+					"Could not evaluate the prerequisite details of feature flag : %v", prereqFeature)
+				return true
+			}
+
+			log.Infof(
+				"Pre requisite flag %v has variation %v for target %v",
+				prereqFeatureConfig.Feature,
+				prereqEvaluatedVariation,
+				target)
+
+			// Compare if the pre requisite variation is a possible valid value of
+			// the pre requisite FF
+			validPrereqVariations := pre.Variations
+			log.Infof(
+				"Pre requisite flag %v should have the variations %v",
+				prereqFeatureConfig.Feature,
+				validPrereqVariations)
+			if !contains(validPrereqVariations, prereqEvaluatedVariation.Identifier) {
+				return false
+			}
+
+			if !e.checkPreRequisite(&prereqFeatureConfig, target) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (e Evaluator) evaluate(identifier string, target *rest.Target, kind string) (rest.Variation, error) {
+	flag, err := e.query.GetFlag(identifier)
+	if err != nil {
+		return rest.Variation{}, err
+	}
+
+	if flag.Kind != kind {
+		return rest.Variation{}, fmt.Errorf("%w, expected: %s, got: %s", ErrFlagKindMismatch, kind, flag.Kind)
+	}
+
+	if flag.Prerequisites != nil {
+		prereq := e.checkPreRequisite(&flag, target)
+		if !prereq {
+			return findVariation(flag.Variations, flag.OffVariation)
+		}
+	}
+	variation, err := e.evaluateFlag(flag, target)
+	if err != nil {
+		return rest.Variation{}, err
+	}
+	return variation, nil
+}
+
+// BoolVariation returns boolean evaluation for target
+func (e Evaluator) BoolVariation(identifier string, target *rest.Target, defaultValue bool) bool {
+	variation, err := e.evaluate(identifier, target, "boolean")
+	if err != nil {
+		// log
```

Comment:
Did you intend to add in log messages here ?

Reply:
Yep